File: books-practice/Data Structures and  Algorithms in Java/src/chap_3/arrays/ds/java/ScoreBoard.java
package chap_3.arrays.ds.java;

/**
 * {@link ScoreBoard} store new entries
 */
public class ScoreBoard {
  // 0
  private int numEntries;
  private GameEntry[] board;

  public ScoreBoard(int capacity) {
    this.board = new GameEntry[capacity];
  }

  public GameEntry getByIndex(int index) {
    return board[index];
  }

  public void add(GameEntry e) {
    int newScore = e.getScore();

    if (numEntries < board.length || newScore > board[numEntries - 1].getScore()) {
      if (numEntries < board.length) {
        numEntries++;
        int j = numEntries - 1;

        while (j > 0 && board[j - 1].getScore() < newScore) {
          board[j] = board[j - 1];
          j--;
        }
        board[j] = e;
      }
    }
  }

  public void add2(GameEntry e) {
    if (numEntries < board.length || e.getScore() > board[numEntries - 1].getScore()) {

      int i = numEntries - 1;

      while (i >= 0 && board[i].getScore() < e.getScore()) {
        if (i + 1 < board.length) {
          board[i + 1] = board[i];
        }
        i--;
      }

      board[i + 1] = e;

      if (numEntries < board.length) {
        numEntries++;
      }
    }
  }

  public GameEntry deleteByIndex(int index) {
    GameEntry solve = board[index];

    if (index < numEntries - 1) {
      int j = index;

      while (j < numEntries-1) {
        board[j] = board[j + 1];
        j++;
      }

      board[numEntries-1] = null;
    }

    return solve;
  }
}

File: books-practice/Data Structures and  Algorithms in Java/src/chap_3/arrays/ds/javascript/ScoreBoard.ts
/**
 * {@link GameEntry} class to store high score entry
 * @param score high score of a user
 * @param name name of the user
 */
export class GameEntry {
  private score: number;
  private name: string;

  constructor(score: number, name: string) {
    this.score = score;
    this.name = name;
  }

  public getScore() {
    return this.score;
  }

  public getName() {
    return this.name;
  }

  public setScore(newScore: number) {
    this.score = newScore;
  }
}

export class ScoreBoard {
  private numEntries: number;
  private arr: Array<GameEntry>;

  constructor(capatity: number) {
    this.arr = new Array(capatity);
    this.numEntries = 0;
  }

  public getTotal(): number {
    return this.arr.length;
  }


  public getByIndex(index: number): GameEntry {
    return this.arr[index];
  }

  public add(entry: GameEntry): void {
    let newEntry = entry.getScore();

    if(this.numEntries < this.arr.length || newEntry > this.arr[this.numEntries-1].getScore()){
      
    }

  }

  public add2(entry: GameEntry): void {

    if (this.numEntries === this.arr.length && entry.getScore() <= this.arr[this.numEntries - 1].getScore()) {
      return;
    }

    let j = 0;
    let pointer = entry;

    while (j < this.arr.length) {
      if (this.arr[j].getScore() < pointer.getScore()) {
        this.arr[j] = pointer;
        pointer = this.arr[j + 1];
      }
      j++;
    }
    this.numEntries++;
  }

  public deletedByIndex(index: number) {
    let j = index;


    while (j < this.numEntries) {

      if (this.arr[j + 1] == null) {
        continue;
      }
      this.arr[j] = this.arr[j + 1];
      j++;
    }
  }
}


const test = new ScoreBoard(5);
const gameEntry1 = new GameEntry(12, "jordan");
const gameEntry2 = new GameEntry(40, "jordan2");
const gameEntry3 = new GameEntry(30, "jordan3");
const gameEntry4 = new GameEntry(100, "jordan3www");


test.add(gameEntry1);
test.add(gameEntry2);
test.add(gameEntry3);
test.add(gameEntry4)

console.log(test.getByIndex(0).getName());
console.log(test.getByIndex(0).getName());
console.log(test.getByIndex(0).getName());
console.log(test.getByIndex(0).getName());